src/coreclr/jit/importer_vectorization.cpp

```diff
@@ -0,0 +1,741 @@
+        simdSize = 32;
+        simdType = TYP_SIMD32;
+
+        niZero   = NI_Vector256_get_Zero;
+        niEquals = NI_Vector256_op_Equality;
+        niCreate = NI_Vector256_Create;
+
+        // Special case: use a single vector for Length == 16
+        useSingleVector = len == 16;
+
+        assert(sizeof(ssize_t) == 8); // this code is guarded with TARGET_64BIT
+        GenTree* long1 = gtNewIconNode(*(ssize_t*)(cns + 0), TYP_LONG);
+        GenTree* long2 = gtNewIconNode(*(ssize_t*)(cns + 4), TYP_LONG);
+        GenTree* long3 = gtNewIconNode(*(ssize_t*)(cns + 8), TYP_LONG);
+        GenTree* long4 = gtNewIconNode(*(ssize_t*)(cns + 12), TYP_LONG);
+        cnsVec1        = gtNewSimdHWIntrinsicNode(simdType, long1, long2, long3, long4, niCreate, baseType, simdSize);
+
+        // cnsVec2 most likely overlaps with cnsVec1:
+        GenTree* long5 = gtNewIconNode(*(ssize_t*)(cns + len - 16), TYP_LONG);
+        GenTree* long6 = gtNewIconNode(*(ssize_t*)(cns + len - 12), TYP_LONG);
+        GenTree* long7 = gtNewIconNode(*(ssize_t*)(cns + len - 8), TYP_LONG);
+        GenTree* long8 = gtNewIconNode(*(ssize_t*)(cns + len - 4), TYP_LONG);
+        cnsVec2        = gtNewSimdHWIntrinsicNode(simdType, long5, long6, long7, long8, niCreate, baseType, simdSize);
+    }
+    else
+#endif
+        if (len <= 16)
+    {
+        // Handle [8..16] inputs via two Vector256
+        assert(len >= 8 && len <= 16);
+
+        simdSize = 16;
+        simdType = TYP_SIMD16;
+
+        niZero   = NI_Vector128_get_Zero;
+        niEquals = NI_Vector128_op_Equality;
+        niCreate = NI_Vector128_Create;
+
+        // Special case: use a single vector for Length == 8
+        useSingleVector = len == 8;
+
+        assert(sizeof(ssize_t) == 8); // this code is guarded with TARGET_64BIT
+        GenTree* long1 = gtNewIconNode(*(ssize_t*)(cns + 0), TYP_LONG);
+        GenTree* long2 = gtNewIconNode(*(ssize_t*)(cns + 4), TYP_LONG);
+        cnsVec1        = gtNewSimdHWIntrinsicNode(simdType, long1, long2, niCreate, baseType, simdSize);
+
+        // cnsVec2 most likely overlaps with cnsVec1:
+        GenTree* long3 = gtNewIconNode(*(ssize_t*)(cns + len - 8), TYP_LONG);
+        GenTree* long4 = gtNewIconNode(*(ssize_t*)(cns + len - 4), TYP_LONG);
+        cnsVec2        = gtNewSimdHWIntrinsicNode(simdType, long3, long4, niCreate, baseType, simdSize);
+    }
+    else
+    {
+        JITDUMP("impExpandHalfConstEqualsSIMD: No V256 support and data is too big for V128\n");
+        // NOTE: We might consider using four V128 for ARM64
+        return nullptr;
+    }
+
+    GenTree* zero = gtNewSimdHWIntrinsicNode(simdType, niZero, baseType, simdSize);
+
+    GenTree* offset1  = gtNewIconNode(dataOffset, TYP_I_IMPL);
+    GenTree* offset2  = gtNewIconNode(dataOffset + len * sizeof(USHORT) - simdSize, TYP_I_IMPL);
+    GenTree* dataPtr1 = gtNewOperNode(GT_ADD, TYP_BYREF, data, offset1);
+    GenTree* dataPtr2 = gtNewOperNode(GT_ADD, TYP_BYREF, gtClone(data), offset2);
```

Comment:
Do we know `data` is cheap to copy? Seems like we need more logic here to spill if the tree is costly.

Also `gtClone` can fail.

Reply:
Seems like perhaps you are relying on callers to ensure these `data` trees are cheap and clonable. If so can you add appropriate comments and/or assertion checks?

Reply:
> Seems like perhaps you are relying on callers to ensure these data trees are cheap and clonable.

Exactly! I'll add asserts